File: Java/MBTITest.java
public class MBTITest{
	
	static String[][] questions = {
		{"Expend energy, enjoy groups", "Conserve energy, enjoy one-on-one"},
		{"Interpret literally", "Look for meaning and possibilities"},
                {"Logical, thinking, questioning", "Empathetic, feeling, accommodating"},
	        {"Organized, orderly", "Flexible, adaptable"},
		{"More outgoing, think out loud", "More reserved, think to yourself"},
		{"Practical, realistic, frank", "Tactful, kind, encouraging"},
	        {"Candid, straightforward, frank", "Tactful, kind, encouraging"},
	        {"Plan, schedule", "Unplanned, spontaneous"},
	        {"Seek many tasks, public activities", "Seek private, solitary activities"},
	        {"Standard, usual, convention", "Different, novel, unique"},
	        {"Firm, tend to criticize, hold the line", "Gentle, tend to appreciate, conciliate"},
	        {"Regulated, structured", "Easy-going, live and let live"},
	        {"External, communicative, express yourself", "Internal, reticent, keep to yourself"},
	        {"Focus on here-and-now", "Look to the future, big picture"},
	        {"Tough-minded, just", "Tender-hearted, merciful"},
	        {"Preparation, plan ahead", "Go with the flow, adapt as you go"},
	        {"Active, initiate", "Reflective, deliberate"},
	        {"Facts, things, what is", "Ideas, dreams, what could be, philosophical"},
	        {"Matter of fact, issue-oriented", "Sensitive, people-oriented, compassionate"},
	{"Control, govern", "Latitude, freedom"}
    };
	static char[] traitForA = {
		'E', 'S', 'T', 'J', 'E', 'S', 'T', 'J', 'E', 'S',
		'T', 'J', 'E', 'S', 'T', 'J', 'E', 'S', 'T', 'J'
    };
	static char[] traitForB = {
        'I', 'N', 'F', 'P', 'I', 'N', 'F', 'P', 'I', 'N',
        'F', 'P', 'I', 'N', 'F', 'P', 'I', 'N', 'F', 'P'
    };
	public static String processAnswers(String answers) {
        if (answers.length() != questions.length) {
	throw new IllegalArgumentException("Answers length must be " + questions.length);
        }
	int[] counts = new int[8];
	for (int i = 0; i < questions.length; i++) {
            char answer = Character.toUpperCase(answers.charAt(i));
            if (answer != 'A' && answer != 'B') {
                throw new IllegalArgumentException("Invalid answer at position " + (i+1) + ". Only A or B allowed.");
            }
            updateTraitCount(answer, i, counts);
        }
	return determineMBTIType(counts);
    }
	private static void updateTraitCount(char answer, int questionIndex, int[] counts) {
        char trait = (answer == 'A') ? traitForA[questionIndex] : traitForB[questionIndex];
        switch (trait) {
            case 'E' -> counts[0]++;
            case 'I' -> counts[1]++;
            case 'S' -> counts[2]++;
            case 'N' -> counts[3]++;
            case 'T' -> counts[4]++;
            case 'F' -> counts[5]++;
            case 'J' -> counts[6]++;
            case 'P' -> counts[7]++;
        }
    }
	private static String determineMBTIType(int[] counts) {
        char first = counts[0] >= counts[1] ? 'E' : 'I';
        char second = counts[2] >= counts[3] ? 'S' : 'N';
        char third = counts[4] >= counts[5] ? 'T' : 'F';
        char fourth = counts[6] >= counts[7] ? 'J' : 'P';
        return "" + first + second + third + fourth;
    }
	public static void displayPersonalityDescription(String type) {
        switch (type) {
            case "INTJ" -> System.out.println("The Architect: Strategic and imaginative thinkers.");
            case "INTP" -> System.out.println("The Thinker: Innovative and curious.");
            case "ENTJ" -> System.out.println("The Commander: Bold, imaginative leaders.");
            case "ENTP" -> System.out.println("The Debater: Clever and intellectual.");
            case "INFJ" -> System.out.println("The Advocate: Quiet, insightful idealists.");
            case "INFP" -> System.out.println("The Mediator: Gentle, empathetic dreamers.");
            case "ENFJ" -> System.out.println("The Protagonist: Charismatic and inspiring.");
            case "ENFP" -> System.out.println("The Campaigner: Energetic, free-spirited.");
            case "ISTJ" -> System.out.println("The Logistician: Responsible and honest.");
            case "ISFJ" -> System.out.println("The Defender: Protective and warm.");
	 case "ESTJ" -> System.out.println("The Executive: Organized and confident.");
            case "ESFJ" -> System.out.println("The Consul: Caring and social.");
            case "ISTP" -> System.out.println("The Virtuoso: Bold and efficient.");
            case "ISFP" -> System.out.println("The Adventurer: Creative and spontaneous.");
            case "ESTP" -> System.out.println("The Entrepreneur: Energetic problem-solver.");
            case "ESFP" -> System.out.println("The Entertainer: Fun-loving and enthusiastic.");
            default -> System.out.println("No description available for this personality type.");
        }
    }
}

File: python/mbtitest.py
QUESTIONS = [
	("Expend energy, enjoy groups", "Conserve energy, enjoy one-on-one"),
	("Interpret literally", "Look for meaning and possibilities"),
	("Logical, thinking, questioning", "Empathetic, feeling, accommodating"),
	("Organized, orderly", "Flexible, adaptable"),
	("More outgoing, think out loud", "More reserved, think to yourself"),
	("Practical, realistic, frank", "Tactful, kind, encouraging"),
	("Candid, straightforward, frank", "Tactful, kind, encouraging"),
	("Plan, schedule", "Unplanned, spontaneous"),
	("Seek many tasks, public activities", "Seek private, solitary activities"),
	("Standard, usual, convention", "Different, novel, unique"),
	("Firm, tend to criticize, hold the line", "Gentle, tend to appreciate, conciliate"),
	("Regulated, structured", "Easy-going, live and let live"),
	("External, communicative, express yourself", "Internal, reticent, keep to yourself"),
	("Focus on here-and-now", "Look to the future, big picture"),
	("Tough-minded, just", "Tender-hearted, merciful"),
	("Preparation, plan ahead", "Go with the flow, adapt as you go"),
	("Active, initiate", "Reflective, deliberate"),
	("Facts, things, what is", "Ideas, dreams, what could be, philosophical"),
	("Matter of fact, issue-oriented", "Sensitive, people-oriented, compassionate"),
	("Control, govern", "Latitude, freedom")
]

TRAITS = [
	('E', 'I'), ('S', 'N'), ('T', 'F'), ('J', 'P'),
        ('E', 'I'), ('S', 'N'), ('T', 'F'), ('J', 'P'),
	('E', 'I'), ('S', 'N'), ('T', 'F'), ('J', 'P'),
        ('E', 'I'), ('S', 'N'), ('T', 'F'), ('J', 'P'),
        ('E', 'I'), ('S', 'N'), ('T', 'F'), ('J', 'P')
]

def process_answers(answers):
    if len(answers) != len(QUESTIONS):
        raise ValueError("Answers must be exactly 20 characters long.")

    counts = {'E': 0, 'I': 0, 'S': 0, 'N': 0, 'T': 0, 'F': 0, 'J': 0, 'P': 0}

    for i, ch in enumerate(answers.upper()):
        if ch not in ('A', 'B'):
            raise ValueError("Invalid answer at position {}: {}".format(i + 1, ch))

        trait = TRAITS[i][0] if ch == 'A' else TRAITS[i][1]
        counts[trait] += 1
    result = ""
    result += 'E' if counts['E'] >= counts['I'] else 'I'
    result += 'S' if counts['S'] >= counts['N'] else 'N'
    result += 'T' if counts['T'] >= counts['F'] else 'F'
    result += 'J' if counts['J'] >= counts['P'] else 'P'

    return result
def describe_INTJ():
    return "The Architect: Strategic and imaginative thinkers."

def describe_INTP():
    return "The Thinker: Innovative and curious."

def describe_ENTJ():
    return "The Commander: Bold, imaginative leaders."

def describe_ENTP():
    return "The Debater: Clever and intellectual."

def describe_INFJ():
    return "The Advocate: Quiet, insightful idealists."

def describe_INFP():
    return "The Mediator: Gentle, empathetic dreamers."

def describe_ENFJ():
    return "The Protagonist: Charismatic and inspiring."

def describe_ENFP():
    return "The Campaigner: Energetic, free-spirited."

def describe_ISTJ():
    return "The Logistician: Responsible and honest."

def describe_ISFJ():
    return "The Defender: Protective and warm."

def describe_ESTJ():
    return "The Executive: Organized and confident."

def describe_ESFJ():
    return "The Consul: Caring and social."

def describe_ISTP():
    return "The Virtuoso: Bold and efficient."

def describe_ISFP():
    return "The Adventurer: Creative and spontaneous."

def describe_ESTP():
    return "The Entrepreneur: Energetic problem-solver."

def describe_ESFP():
    return "The Entertainer: Fun-loving and enthusiastic."

def describe_default():
    return "No description available."
def get_description(personality_type):
    if personality_type == "INTJ":
        return describe_INTJ()
    elif personality_type == "INTP":
        return describe_INTP()
    elif personality_type == "ENTJ":
        return describe_ENTJ()
    elif personality_type == "ENTP":
        return describe_ENTP()
    elif personality_type == "INFJ":
        return describe_INFJ()
    elif personality_type == "INFP":
        return describe_INFP()
    elif personality_type == "ENFJ":
        return describe_ENFJ()
    elif personality_type == "ENFP":
        return describe_ENFP()
    elif personality_type == "ISTJ":
        return describe_ISTJ()
    elif personality_type == "ISFJ":
        return describe_ISFJ()
    elif personality_type == "ESTJ":
        return describe_ESTJ()
    elif personality_type == "ESFJ":
        return describe_ESFJ()
    elif personality_type == "ISTP":
        return describe_ISTP()
    elif personality_type == "ISFP":
        return describe_ISFP()
    elif personality_type == "ESTP":
        return describe_ESTP()
    elif personality_type == "ESFP":
        return describe_ESFP()
    else:
        return describe_default()
